### PR TITLE
Use spans for serialization

### PR DESCRIPTION
**Pull request by paulomorgado**
Addresses #6543

## Changes

Refactored serialization logic to use `Span<byte>` for buffer manipulation, improving performance, memory efficiency, and code readability. Updated method signatures to eliminate explicit `writePosition` management, replaced incremental write logic with consistent `+=` operations, and streamlined encoding methods. Enhanced test coverage to validate edge cases and updated all test cases to align with the new implementation. Applied changes consistently across all serializers, ensuring backward compatibility and laying the groundwork for future optimizations. Improved error handling, debugging, and documentation for better maintainability and robustness.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by paulomorgado** [comment]
Hi @martincostello, @Kielek, can you help me here with the test failures?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpLogSerializer.cs [hunk 1]
```diff
@@ -148,14 +147,13 @@ internal static int WriteScopeLog(byte[] buffer, int writePosition, SdkLimitOpti
...
-        writePosition = ProtobufSerializer.WriteTagAndLength(buffer, writePosition, numberOfUtf8CharsInString + 1 + serializedLengthSize, ProtobufOtlpLogFieldNumberConstants.ScopeLogs_Scope, ProtobufWireType.LEN);
-        writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpCommonFieldNumberConstants.InstrumentationScope_Name, numberOfUtf8CharsInString, value);
+        writePosition += ProtobufSerializer.WriteTagAndLength(buffer.AsSpan(writePosition), numberOfUtf8CharsInString + 1 + serializedLengthSize, ProtobufOtlpLogFieldNumberConstants.ScopeLogs_Scope, ProtobufWireType.LEN);
+        writePosition += ProtobufSerializer.WriteStringWithTag(buffer.AsSpan(writePosition), ProtobufOtlpCommonFieldNumberConstants.InstrumentationScope_Name, numberOfUtf8CharsInString, value);
 
         for (int i = 0; i < logRecords.Count; i++)
         {
             writePosition = WriteLogRecord(buffer, writePosition, sdkLimitOptions, experimentalOptions, logRecords[i]);
         }
-
```
Comment (martincostello) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpLogSerializer.cs:158]: Put the blank line back, and the style analyser won't fail the build.

Comment (Kielek) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpLogSerializer.cs:158]: In general, it is good practice to run local build in `Release` mode. `Debug` allows to compile with Warnings, Release is extremely strict.

Comment (paulomorgado) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpLogSerializer.cs:158]: My bad! Sorry!

I have fixed the style errors, but I still have a test failing:

```
Failed OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.OtlpAttributeTests.IntegralTypesSupported(value: [1, 2, 3]) [< 1 ms]
Error Message:
 Google.Protobuf.InvalidProtocolBufferException : Protocol message contained an invalid tag (zero).
```

I can't figure out how this is related to my changes.

Runing the tests in debug mode I get this:

```
 OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests.OtlpAttributeTests.IntegralTypesSupported(value: [1, 2, 3])
   Source: OtlpAttributeTests.cs line 56
   Duration: 7 ms

  Message: 
Microsoft.VisualStudio.TestPlatform.TestHost.DebugAssertException : Method Debug.Fail failed with 'bytesWritten did not match numberOfUtf8CharsInString
', and was translated to Microsoft.VisualStudio.TestPlatform.TestHost.DebugAssertException to avoid terminating the process hosting the test.

  Stack Trace: 
TestHostTraceListener.GetException(String message)
TestHostTraceListener.Fail(String message, String detailMessage)
TraceInternal.Fail(String message, String detailMessage)
Debug.Fail(String message, String detailMessage)
ProtobufOtlpTagWriter.TryWriteByteArrayTag(OtlpTagWriterState& state, String key, ReadOnlySpan`1 value) line 94
TagWriter`2.TryWriteTag(TTagState& state, String key, Object value, Nullable`1 tagValueMaxLength) line 70
TagWriter`2.TryWriteTag(TTagState& state, KeyValuePair`2 tag, Nullable`1 tagValueMaxLength) line 28
OtlpAttributeTests.TryTransformTag(KeyValuePair`2 tag, KeyValue& attribute) line 293
OtlpAttributeTests.IntegralTypesSupported(Object value) line 59
```

This happens because `ProtobufOtlpTagWriter.TryWriteByteArrayTag` is calling `ProtobufSerializer.WriteStringWithTag` with `ProtobufOtlpCommonFieldNumberConstants.KeyValue_Key` (1) for `numberOfUtf8CharsInString` with `key` as `"key"`.

What am I missing here?

**Comment by martincostello** [comment]
It would be good to get some before and after benchmarks on the changes here.

Looking at the diff I have a slight concern that API has changed in a way that makes it harder to use/easier to make a mistake (e.g. use `=` instead of `+=`).

Numbers on the performance difference the change makes would help guide whether the gain is worth it if there's no way to refactor to make it easier to not misuse the methods (e.g. maybe remove the return value and pass the write position as a `ref`?).

**Comment by paulomorgado** [comment]
> It would be good to get some before and after benchmarks on the changes here.
> 
> Looking at the diff I have a slight concern that API has changed in a way that makes it harder to use/easier to make a mistake (e.g. use `=` instead of `+=`).
> 
> Numbers on the performance difference the change makes would help guide whether the gain is worth it if there's no way to refactor to make it easier to not misuse the methods (e.g. maybe remove the return value and pass the write position as a `ref`?).

That's why I changed (I hope everywhere) the variable/parameter name from `writePosition` to `bytesWritten`.

This should probably be better with a `ref Span<byte>` and returning bytes written to the caller to use if needed.

There is still an issue with state structs with a `byte[]` parameter that cannot be a field. It needs to be a method parameter or a `Memory<byte>` and I don't want to do the latter.

**Comment by paulomorgado** [comment]
Superseded by #7451
